pallets/parachain-staking/src/lib.rs

```diff
@@ -403,6 +403,31 @@ pub mod pallet {
 		type MinNominatorStk: Get<BalanceOf<Self>>;
 		/// Weight information for extrinsics in this pallet.
 		type WeightInfo: WeightInfo;
+
+		//Philosophy about this feature:
+		// An attacker who intends to skip slots can execute their attack whether with or without a
+		// preliminary check. The utility of this check is that it prevents stakers from
+		// accidentally forgetting to set their session keys. There is cost associated with this
+		// check. At minimum we call this function (which in moonbeam's case does a storage lookup)
+		// for the target number of candidates in every round. If some don't meet the prereqs, then
+		// we call it for even more.
+		//
+		// Because the election is for the good of the chain rather than triggered by any one account
+		// the cost of this election is born by the entire network in the form of less weight
+		// available in the block.
+		//
+		// So effectively this check protects collators who are not prepared to register their
+		// session keys from missed rewards (and in the future possibly slashes), at the expense of
+		// the entire network.
+		//
+		// I believe this is undesirable given our already-aligned economic incentives. Collators
+		// who don't do their elected duties (including their duty to register their session keys)
+		// are supposed to be punished and doing so incentives them to be more careful. If they
+		// continue to fail at their duties, they will lose nominations and naturally not be elected.
```

Comment:
Here's my assessment of what this feature does and an argument for why we don't want it. Feedback welcome.

Contextually, I'm worried that his feature is mainly a reaction to yesterday's stall. The stall was caused by a combination sudden drastic nomination changes that are not realistic in a public network and a bug with the --collator flag that is fixed in #455.

Reply:
For me this is more of a political question than a technical one.

What precautions should the dev team take to handle the case where the voting token holders elect a collator set, none of whom fulfill the responsibility to register their session keys. One reasonable answer is that the whole point of token voting is to let the token holders decide who their collators are. If the voters want to elect collators who don't get their session keys registered, we better let them. In practice, I won't be nominating those candidates.

Or to generalize a little bit, it could be phrased as: what precautions should the dev team take to handle the case where the token holders elect a collator set of whom the dev team does not approve?